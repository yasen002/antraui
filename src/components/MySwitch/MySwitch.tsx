import { FC, MouseEvent, useState, useRef, useEffect } from 'react';

type SwitchColor = "primary" | "secondary" | "warning" | "default";

type SwitchSize = "small" | "medium";

interface IMySwitchProps {
    /**
    * customize the color of the switch
    */
    color?: SwitchColor,
    /**
    * customize the size of the switch
    */
    size?: SwitchSize,
    /**
    * customize the label of switch
    */
    label?: string,
    /**
    * used to disable the switch
    */
    disabled?: boolean,
    /**
    * used to check the button
    */
    checked?: boolean
    /**
    * event handler for the click event of the switch
    */
    onChange?: (event: MouseEvent) => void
}

const MySwitch: FC<IMySwitchProps> = ({
    color = "primary",
    size = "medium",
    label,
    disabled = false,
    checked = false,
    onChange
}) => {

    const [isChecked, setIsChecked] = useState(checked);
    
    const thumbRef = useRef<HTMLDivElement>(null);

    useEffect(() => {
        setIsChecked(checked);
    }, [checked])

    const handleClick = (e: MouseEvent) => {
        if (disabled) return;
        setIsChecked(!isChecked);
        // Add animation after click
        if (thumbRef && thumbRef.current) {
            thumbRef.current.style.animation = !isChecked ? `${color}-${size}-effect 0.3s linear` : `Off-${size}-effect 0.3s linear`;
        }
        onChange?.(e);
    }


    const switchContainer: string = ["switch", disabled ? `switch-disabled` : ""].join(" ");
    const switchStyle: string = isChecked ? `switch-${color}-${size}` : `switch-${size}-off`;

    return (
        <div data-testid="MySwitch--container" className={switchContainer} onClick={handleClick}>
            <div data-testid="MySwitch--switch" className={switchStyle} >
                <div className='switch-track-wrapper'><div className='switch-track'></div></div>
                <div className='switch-thumb-wrapper'><div data-testid="switch-thumb" className='switch-thumb' ref={thumbRef}></div></div>
            </div>
            {label}
        </div>
    )
}


export default MySwitch;
