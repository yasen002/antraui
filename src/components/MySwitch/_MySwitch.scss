@use "../../scss/variable";

$switch-light-color: #FAFAFA;
$switch-dark-color: #000000;
$switch-disabled-color: #BDBDBD;

// The common css style
.switch {
    font-family: variable.$font-family;
    font-style: normal;
    font-weight: 400;
    font-size: 16px;
    line-height: 150%;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
    font-size: 16px;
    line-height: 24px;
    letter-spacing: 0.15px;
}

.switch-track-wrapper {
    box-sizing: border-box;
    width: 100%;
    height: 100%;
}

.switch-thumb-wrapper {
    box-sizing: border-box;
    position: absolute;
    top: 0;
    left: 0;
    transition: transform 0.2s ease-out;
}

.switch-thumb {
    box-shadow: 0px 2px 1px rgba($switch-dark-color, 0.2), 0px 1px 1px rgba($switch-dark-color, 0.14), 0px 1px 3px rgba($switch-dark-color, 0.12);
}

// switch-color-map
$color-map: (
    primary: (default: #3F51B5,
        hover: rgba(63, 81, 181, 0.08)),
    secondary: (default: #F50057,
        hover: rgba(245, 0, 87, 0.08)),
    warning: (default: #ed6c02,
        hover: rgba(#ed6c02, 0.08)),
    default: (default: $switch-light-color,
        hover: rgba($switch-dark-color, 0.04)),
);

@mixin smallSize {
    width: 40px;
    height: 24px;
    position: relative;

    .switch-track-wrapper {
        padding: 7px;

        .switch-track {
            width: 26px;
            height: 10px;
            border-radius: 5px;
            opacity: 0.5;
        }
    }

    .switch-thumb-wrapper {
        width: 24px;
        height: 24px;
        padding: 4px;
        border-radius: 19px;

        .switch-thumb {
            width: 16px;
            height: 16px;
            border-radius: 10px;
        }
    }
}

@mixin mediumSize {
    width: 58px;
    height: 38px;
    position: relative;

    .switch-track-wrapper {
        padding: 12px;

        .switch-track {
            width: 34px;
            height: 14px;
            border-radius: 7px;
            opacity: 0.5;
        }
    }

    .switch-thumb-wrapper {
        width: 38px;
        height: 38px;
        padding: 9px;
        border-radius: 19px;

        .switch-thumb {
            width: 20px;
            height: 20px;
            border-radius: 10px;
        }
    }
}

// Generate style for checked color switch
@mixin color-checked-class($color, $size) {
    .switch-#{$color}-#{$size} {
        @if $size =="medium" {
            @include mediumSize;

            .switch-thumb-wrapper {
                transform: translateX(20px);
            }
        }

        @else {
            @include smallSize;

            .switch-thumb-wrapper {
                transform: translateX(16px);
            }
        }

        @if $color =='default' {
            .switch-track {
                background-color: $switch-dark-color;
            }
        }

        @else {
            .switch-track {
                background-color: map-get(map-get($color-map, $color), "default");

            }
        }

        .switch-thumb-wrapper .switch-thumb {
            background-color: map-get(map-get($color-map, $color), "default");
        }

        &:hover {
            .switch-thumb-wrapper {
                background-color: map-get(map-get($color-map, $color), "hover");
            }
        }
    }
}

// Generate style for off switch
@mixin switchOff($size) {
    .switch-#{$size}-off {
        @if $size =="medium" {
            @include mediumSize;
        }

        @else {
            @include smallSize;

        }

        .switch-track {
            background-color: $switch-dark-color;
            opacity: 0.38;

        }

        .switch-thumb {
            background-color: $switch-light-color;
        }

        &:hover {
            .switch-thumb-wrapper {
                background-color: rgba($switch-dark-color, 0.04);
            }
        }
    }
}

$colors: (
    "primary",
    "secondary",
    "warning",
    "default"
);
$size: (
    "small",
    "medium"
);

// Generate animations for switch change
@mixin color-keyframes($color, $size) {
    @if $size =='medium' {
        @keyframes #{$color}-#{$size}-effect {
            0% {
                outline: 2px solid rgba(map-get(map-get($color-map, $color), "default"), 0.4);
            }

            50% {
                outline: 5px solid rgba(map-get(map-get($color-map, $color), "default"), 0.2);
            }

            100% {
                outline: 9px solid rgba(map-get(map-get($color-map, $color), "default"), 0);
            }
        }
    }

    @else {
        @keyframes #{$color}-#{$size}-effect {
            0% {
                outline: 1px solid rgba(map-get(map-get($color-map, $color), "default"), 0.4);
            }

            50% {
                outline: 2px solid rgba(map-get(map-get($color-map, $color), "default"), 0.2);
            }

            100% {
                outline: 4px solid rgba(map-get(map-get($color-map, $color), "default"), 0);
            }
        }
    }
}

@keyframes Off-medium-effect {
    0% {
        outline: 2px solid rgba($switch-light-color, 0.4);
    }

    50% {
        outline: 5px solid rgba($switch-light-color, 0.2);
    }

    100% {
        outline: 9px solid rgba($switch-light-color, 0);
    }
}

@keyframes Off-small-effect {
    0% {
        outline: 1px solid rgba($switch-light-color, 0.3);
    }

    50% {
        outline: 2px solid rgba($switch-light-color, 0.1);
    }

    100% {
        outline: 4px solid rgba($switch-light-color, 0);
    }
}


@mixin generate-color-checked-classes {
    @each $size in $size {
        @include switchOff($size);

        @each $color in $colors {
            @include color-checked-class($color, $size);
            @include color-keyframes($color, $size);
        }
    }
}

@include generate-color-checked-classes();


.switch-disabled {
    pointer-events: none;
    cursor: none;
    color: rgba($switch-dark-color, 0.36);

    .switch-track {
        background-color: rgba($switch-dark-color, 0.36);
    }

    .switch-thumb-wrapper .switch-thumb {
        background-color: $switch-disabled-color;
    }
}