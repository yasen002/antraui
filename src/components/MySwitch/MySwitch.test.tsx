import { fireEvent, render, screen } from "@testing-library/react";
import MySwitch from "./MySwitch";

describe("MySwitch test cases", () => {
    test("snapshot test for MySwitch", () => {
        const view = render(<MySwitch />);
        expect(view.asFragment()).toMatchSnapshot();
    })

    test("MySwitch component should be rendered", () => {
        render(<MySwitch />);
        const switchEl = screen.queryByTestId("AntraUI-Switch-Container");
        expect(switchEl).toBeInTheDocument();
    })
 
    test("MySwitch should change its label based on label props", () => {
        render(<MySwitch label="Switch" />)
        const switchEl = screen.queryByText("Switch");
        expect(switchEl).toBeInTheDocument();
    })

    test("MySwitch should change its checked based on checked props", async () => {
        const { rerender } = render(<MySwitch checked={true} />);
        let switchEl = await screen.findByTestId("AntraUI-Switch-Container");
        expect(switchEl).toHaveClass("AntraUI-Switch-isChecked");
        rerender(<MySwitch checked={false} />);
        switchEl = await screen.findByTestId("AntraUI-Switch-Container");
        expect(switchEl).toHaveClass("AntraUI-Switch-notChecked");
    })

    test("MySwitch should change its disabled based on disabled props", async () => {
        render(<MySwitch disabled />);
        let switchEl = await screen.findByTestId("AntraUI-Switch-Container");
        expect(switchEl).toHaveClass("AntraUI-Switch-isDisabled");
    })

    test("MySwitch should change its color based on color props", async () => {
        const { rerender } = render(<MySwitch color="primary" checked={true} />);
        let switchEl = await screen.findByTestId("AntraUI-Switch-Container");
        expect(switchEl).toHaveClass("AntraUI-Switch-primary");
        rerender(<MySwitch color="secondary" checked={true} />)
        expect(switchEl).toHaveClass("AntraUI-Switch-secondary");
        rerender(<MySwitch color="default" checked={true} />)
        expect(switchEl).toHaveClass("AntraUI-Switch-default");
    })

    test("MySwitch should change its size based on size props", async () => {
        const { rerender } = render(<MySwitch color="primary" checked={true} />);
        let switchEl = await screen.findByTestId("AntraUI-Switch-Container");
        expect(switchEl).toHaveClass("AntraUI-Switch-medium");
        rerender(<MySwitch color="primary" checked={true} size="small" />)
        expect(switchEl).toHaveClass("AntraUI-Switch-small");
    })

    test("MySwitch should be disabled if the disables props is provided, , should invoke onClick when it is not disabled", async () => {
        const mockClickHandler = jest.fn();
        const { rerender } = render(<MySwitch onChange={mockClickHandler} />);
        let switchEl = await screen.findByTestId("AntraUI-Switch-Container");
        fireEvent.click(switchEl);
        expect(mockClickHandler).toBeCalledTimes(1);
        rerender(<MySwitch onChange={mockClickHandler} disabled/>);
        fireEvent.click(switchEl);
        expect(mockClickHandler).toBeCalledTimes(1);
    })

    test("MySwitch shuld have ripple animation on change switch", async() => {
        render(<MySwitch />);
        let thumbEl = await screen.findByTestId("AntraUI-Switch-Head");
        fireEvent.click(thumbEl);
        let children = await screen.findByTestId("AntraUI-Switch-Ripple");
        expect(children).toBeDefined()
    })
})