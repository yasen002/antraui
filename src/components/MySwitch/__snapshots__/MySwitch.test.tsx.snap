// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MySwitch test cases snapshot test for MySwitch 1`] = `
<DocumentFragment>
  <div
    class="switch "
    data-testid="MySwitch--container"
  >
    <div
      class="switch-medium-off"
      data-testid="MySwitch--switch"
    >
      <div
        class="switch-track-wrapper"
      >
        <div
          class="switch-track"
        />
      </div>
      <div
        class="switch-thumb-wrapper"
      >
        <div
          class="switch-thumb"
          data-testid="switch-thumb"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
